### PR TITLE
feat: add bearer_token authenticator

### PR DESCRIPTION
**Pull request by wezzle**
## Related issue

https://community.ory.sh/t/oathkeeper-example-with-kratos-api-session-token/2373

## Proposed changes

Adds a new authenticator to work with Kratos' new API token.
Works the same as the cookie_session authenticator but checks for a bearer token in the Authorization header (unless overwritten by token_from)

## Checklist

- [x] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md).
- [x] I have read the [security policy](../security/policy).
- [x] I confirm that this pull request does not address a security
      vulnerability. If this pull request addresses a security. vulnerability, I
      confirm that I got green light (please contact
      [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push
      the changes.
- [x] I have added tests that prove my fix is effective or that my feature
      works.
- [x] I have added or changed [the documentation](docs/docs).

## Further comments

If this is the way to go I can add this to the docs as well, just let me know.

### CODE REVIEW

**Review by aeneasr — CHANGES_REQUESTED** [review]
Great, thank you! Could you please add some documentation as well? A good place would be around here, probably: https://github.com/ory/oathkeeper/blob/master/docs/docs/pipeline/authn.md#cookie_session

**Comment by wezzle** [comment]
@aeneasr yes, I'll get right on that.

**Comment by aeneasr** [comment]
Hm yeah maybe just call it `bearer_token`?

**Comment by wezzle** [comment]
Ok, i'll rename it and add it to the docs

**Comment by wezzle** [comment]
@aeneasr the authenticator has been renamed and docs have been added. Please let me know if anything else is needed.

File: pipeline/authn/authenticator_bearer_token.go [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+	if err := a.c.AuthenticatorConfig(a.GetID(), config, &c); err != nil {
+		return nil, NewErrAuthenticatorMisconfigured(a, err)
+	}
+
+	if len(c.ExtraFrom) == 0 {
+		c.ExtraFrom = "extra"
+	}
+
+	if len(c.SubjectFrom) == 0 {
+		c.SubjectFrom = "subject"
```
Comment (aeneasr) [inline pipeline/authn/authenticator_bearer_token.go:68]: OAuth2 uses the shorthand `sub` - I think we should use `sub` per default also:

```suggestion
		c.SubjectFrom = "sub"
```

File: docs/docs/pipeline/authn.md [hunk 2]
```diff
@@ -328,6 +328,129 @@ HTTP/1.0 401 Status Unauthorized
...
+  will be preserved instead of replacing the path with the path of the request
+  being checked.
+- `extra_from` (string, optional - defaults to `extra`) - A
+  [GJSON Path](https://github.com/tidwall/gjson/blob/master/SYNTAX.md) pointing
+  to the `extra` field. This defaults to `extra`, but it could also be `@this`
+  (for the root element), `session.foo.bar` for
+  `{ "subject": "...", "session": { "foo": {"bar": "whatever"} } }`, and so on.
+- `subject_from` (string, optional - defaults to `subject`) - A
+  [GJSON Path](https://github.com/tidwall/gjson/blob/master/SYNTAX.md) pointing
+  to the `subject` field. This defaults to `subject`. Example: `identity.id` for
```
Comment (aeneasr) [inline docs/docs/pipeline/authn.md:352]:
```suggestion
  to the `subject` field. This defaults to `sub`. Example: `identity.id` for
```

**Review by aeneasr — CHANGES_REQUESTED** [review]
Awesome! Just one more little change :)

File: docs/docs/pipeline/authn.md (hunk 2 above)
Comment (wezzle) [reply docs/docs/pipeline/authn.md]: I've just sent in the change. This does make it diverge from the cookie_session configuration, but perhaps that will get the change sometime in the future as well?

File: .schema/config.schema.json [hunk 3]
```diff
@@ -443,6 +443,85 @@
...
+          "title": "Extra JSON Path",
+          "description": "The `extra` field in the ORY Oathkeeper authentication session is set using this JSON Path. Defaults to `extra`, and could be `@this` (for the root element), `foo.bar` (for key foo.bar), or any other valid GJSON path. See [GSJON Syntax](https://github.com/tidwall/gjson/blob/master/SYNTAX.md) for reference.",
+          "type": "string",
+          "default": "extra"
+        },
+        "subject_from": {
+          "title": "Subject JSON Path",
+          "description": "The `subject` field in the ORY Oathkeeper authentication session is set using this JSON Path. Defaults to `subject`. See [GSJON Syntax](https://github.com/tidwall/gjson/blob/master/SYNTAX.md) for reference.",
+          "type": "string",
+          "default": "subject"
```
Comment (aeneasr) [inline .schema/config.schema.json:517]:
```suggestion
          "default": "sub"
```

File: docs/docs/pipeline/authn.md [hunk 4]
```diff
@@ -328,6 +328,129 @@ HTTP/1.0 401 Status Unauthorized
...
+  method/path/headers to for validation.
+- `preserve_path` (boolean, optional) - If set, any path in `check_session_url`
+  will be preserved instead of replacing the path with the path of the request
+  being checked.
+- `extra_from` (string, optional - defaults to `extra`) - A
+  [GJSON Path](https://github.com/tidwall/gjson/blob/master/SYNTAX.md) pointing
+  to the `extra` field. This defaults to `extra`, but it could also be `@this`
+  (for the root element), `session.foo.bar` for
+  `{ "subject": "...", "session": { "foo": {"bar": "whatever"} } }`, and so on.
+- `subject_from` (string, optional - defaults to `subject`) - A
```
Comment (aeneasr) [inline docs/docs/pipeline/authn.md:350]:
```suggestion
- `subject_from` (string, optional - defaults to `sub`) - A
```

File: docs/docs/pipeline/authn.md [hunk 5]
```diff
@@ -328,6 +328,129 @@ HTTP/1.0 401 Status Unauthorized
...
+  [GJSON Path](https://github.com/tidwall/gjson/blob/master/SYNTAX.md) pointing
+  to the `subject` field. This defaults to `sub`. Example: `identity.id` for
```
Comment (aeneasr) [inline docs/docs/pipeline/authn.md:352]:
```suggestion
  to the `sub` field. This defaults to `sub`. Example: `identity.id` for
```

**Pull request merged**